pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +73,97 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Roll processes previous Rollout object and compares to the current
+// rollout state. If there is different, Roll will start or stop the rollout
+// and update `RevisionRollout` objects accoringly.
+// Roll returns true if any changes have been made.
+func (cur *Rollout) Roll(prev *Rollout) bool {
+	if prev == nil {
+		return false
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// Optimize in the later passes.
+	// Map by tag.
+	currConfigs, prevConfigs := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i, cfg := range cur.Configurations {
+		currConfigs[cfg.Tag] = append(currConfigs[cfg.Tag], &cur.Configurations[i])
+	}
+	for i, cfg := range prev.Configurations {
+		prevConfigs[cfg.Tag] = append(prevConfigs[cfg.Tag], &prev.Configurations[i])
+	}
+	ret := false
+	for t, cfgs := range currConfigs {
+		pcfgs, ok := prevConfigs[t]
+		// A new tag was added, so we have no previous state to roll from,
+		// thus just move over, we'll rollout to 100% from the get go (and it is
+		// always 100%, since default tag is _always_ there).
+		if !ok {
+			continue
+		}
+		// This is basically an intersect algorithm,
+		// It relies on the fact that inputs are sorted.
+		for i, j := 0, 0; i < len(cfgs) && j < len(pcfgs); {
+			switch {
+			case cfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
+				// Config might have 0% traffic assigned, if it is a tag only route (i.e.
+				// receives no traffic via default tag).
+				if cfgs[i].Percent != 0 {
+					ret = ret || rollConfig(cfgs[i], pcfgs[j])
+				}
+				i++
+				j++
+			case cfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
+				// A new config, has been added. No action for rollout though.
+				i++
+			default: // cur > prev.
+				// A config has been removed. Again, no action for rollout, since
+				// this will no longer be rolling out even it were.
+				j++
+			}
+		}
+		// Non overlapping configs don't matter since we either roll 100% or remove them.
+	}
+	return ret
+}
+
+func rollConfig(cur, prev *ConfigurationRollout) bool {
```

Comment:
I think it'd be useful to start the invariants about cur/prev here.

cur/prev are the goal/actual state of a rollout for a single configuration of a single tag.

This function alters `cur` to be the appropriate "next step" in this configuration's rollout, and returns whether any changes were made.

Does that sound right?

Reply:
I've rewritten this comment, ptal.

Reply:
This is also altering "prev" in certain ways.

You might hate this, but I'm going to suggest it anyways... 😅 

What if we adopted a more functional approach to this where instead of returning a bool we returned `*ConfigurationRollout`, and left cur/prev untouched.  nil == no work, non-nil is the next step.

I might even suggest we do this for your "Roll" function above (returning a `*Rollout`).